modules/OFConnectionManager2/module/src/bundle.c

```diff
@@ -500,52 +525,52 @@ bundle_task(void *cookie)
 
     while (state->cur_subbundle < state->subbundle_count) {
         subbundle_t *subbundle = &state->subbundles[state->cur_subbundle];
+
         /* iterate through the current subbundle */
         while (state->cur_offset < subbundle->count) {
+            /* cxn is gone or outstanding operations have been done */
             if (cxn) {
-
-                AIM_LOG_TRACE("cur_subbundle=%u bundle_task cur_offset=%u is_pending %d",
-                              state->cur_subbundle,
-                              state->cur_offset, state->cur_offset_is_pending);
-                if (state->cur_offset_is_pending == false) {
-                    /* The task is ok to process message at cur_offset.
-                     * The cur_offset is in clear status. */
-                    of_object_t *obj = subbundle->objs[state->cur_offset];
-                    if (obj) {
-                        indigo_error_t rv;
-                        rv = ind_cxn_process_message(cxn, obj);
-                        if (rv == INDIGO_ERROR_PENDING) {
-                            /* The message is an async operation
-                             * Mark the messsage at cur_offset as pending.
-                             * The pending status will be cleared in the next
-                             * runnable cycle.
-                             * The lower layer decides to block the bundle loop.
-                             */
-                            state->cur_offset_is_pending = true;
-                            AIM_LOG_TRACE("bundle_task cur_offset=%u pending\n",
-                                          state->cur_offset);
-                            return IND_SOC_TASK_CONTINUE;
-                        }
+                AIM_LOG_TRACE("cur_subbundle=%u cur_offset=%u is_paused=%d",
+                              state->cur_subbundle, state->cur_offset,
+                              state->cur_subbundle_is_paused);
+
+                /* The task is ok to process message at cur_offset.
+                 * The cur_offset is in clear status. */
+                of_object_t *obj = subbundle->objs[state->cur_offset];
+                if (obj) {
+                    indigo_error_t rv;
+                    rv = ind_cxn_process_message(cxn, obj);
+                    if (rv == INDIGO_ERROR_PENDING) {
+                        /* The message is an async operation. Set pause flag. */
+                        state->cur_subbundle_is_paused = true; 
+                        AIM_LOG_TRACE("bundle_task cur_offset=%u pending\n",
+                                      state->cur_offset);
                     }
-                } else {
-                    /* This subbundle msg is done. Clear its pending status.
-                     * Fall through to free it.
+                    /* rv is OK, ERROR, or CONTINUE, then allow next message.
+                     * The obj should have been duplicated in the OFStateManager
```

Comment:
now that we have a CONTINUE return code, how is the bundle processing code supposed to treat it?  we have already specified that if the lower level driver returns PENDING, the lower level driver must call the resume API at some later point, so that we can decrement the pending count internally.  what is the expected lower level driver behavior in the case of CONTINUE?  this behavior also needs to be documented in the gentable API comment in modules/indigo/module/inc/indigo/of_state_manager.h.

Reply:
They will be in the new design.